compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java

```diff
@@ -5019,6 +5021,88 @@ public void visit(BLangForeach foreach) {
         }
     }
 
+    private boolean canEliminateIterator(BLangForeach loop) {
+        BLangExpression collection = loop.collection;
+        return switch (collection.getKind()) {
+            case BINARY_EXPR, LIST_CONSTRUCTOR_EXPR -> true;
+            case SIMPLE_VARIABLE_REF -> {
```

Comment:
We can also include function call expressions, let expressions and query expressions.

Reply:
Fixed